napari/utils/colormaps/colormap.py

```diff
@@ -460,48 +556,170 @@ def _zero_preserving_modulo_numpy(
-            everywhere else.
-        """
-        result = np.empty_like(values, dtype=dtype)
+def _cast_direct_labels_to_minimum_type_numpy(
+    data: np.ndarray, direct_colormap: DirectLabelColormap
+) -> np.ndarray:
+    """
+    Cast direct labels to the minimum type.
 
-        for i in numba.prange(values.size):
-            if values.flat[i] == to_zero:
-                result.flat[i] = 0
-            else:
-                result.flat[i] = (values.flat[i] - 1) % n + 1
+    Parameters
+    ----------
+    data : np.ndarray
+        The input data array.
+    direct_colormap : DirectLabelColormap
+        The direct colormap.
 
-        return result
+    Returns
+    -------
+    np.ndarray
+        The casted data array.
+    """
+    max_value = max(x for x in direct_colormap.color_dict if x is not None)
+    if max_value > 2**16:
+        raise RuntimeError(  # pragma: no cover
+            "Cannot use numpy implementation for large values of labels "
+            "direct colormap. Please install numba."
+        )
+    dtype = minimum_dtype_for_labels(direct_colormap.unique_colors_num() + 2)
+    label_mapping = direct_colormap.values_mapping_to_minimum_values_set()[0]
+
+    mapper = np.full((max_value + 2), DEFAULT_VALUE, dtype=dtype)
+    for key, val in label_mapping.items():
+        if key is None:
+            continue
+        mapper[key] = val
+
+    if data.dtype.itemsize > 2:
+        data = np.clip(data, 0, max_value + 1)
+    return mapper[data]
+
+
+def _zero_preserving_modulo_jit(
+    values: np.ndarray, n: int, dtype: np.dtype, to_zero: int = 0
+) -> np.ndarray:
+    """``(values - 1) % n + 1``, but with one specific value mapped to 0.
+
+    This ensures (1) an output value in [0, n] (inclusive), and (2) that
+    no nonzero values in the input are zero in the output, other than the
+    ``to_zero`` value.
+
+    Parameters
+    ----------
+    values : np.ndarray
+        The dividend of the modulo operator.
+    n : int
+        The divisor.
+    dtype : np.dtype
+        The desired dtype for the output array.
+    to_zero : int, optional
+        A specific value to map to 0. (By default, 0 itself.)
+
+    Returns
+    -------
+    np.ndarray
+        The result: 0 for the ``to_zero`` value, ``values % n + 1``
+        everywhere else.
+    """
+    result = np.empty_like(values, dtype=dtype)
+
+    for i in prange(values.size):
+        if values.flat[i] == to_zero:
+            result.flat[i] = 0
+        else:
+            result.flat[i] = (values.flat[i] - 1) % n + 1
+
+    return result
+
+
+def _cast_direct_labels_to_minimum_type_for_jit(
+    data: np.ndarray, direct_colormap: DirectLabelColormap
+) -> np.ndarray:
+    """
+    Cast direct labels to the minimum type.
+
+    Parameters
+    ----------
+    data : np.ndarray
+        The input data array.
+    direct_colormap : DirectLabelColormap
+        The direct colormap.
+
+    Returns
+    -------
+    np.ndarray
+        The casted data array.
```

Comment:
```suggestion
        The cast data array.
```

(Cast is an irregular verb — the past participle [is also "cast"](https://www.englisch-hilfen.de/en/grammar/unreg_verben1.htm).)